drivers/regulator/regulator_pca9422.c

```diff
@@ -0,0 +1,1234 @@
+	.enmode_pos = PCA9422_LDO3_CFG_L3_ENMODE_POS,
+	.lpmode_mask = PCA9422_LDO3_CFG_L3_LPMODE_MASK,
+	.lpmode_pos = PCA9422_LDO3_CFG_L3_LPMODE_POS,
+
+	.ad_reg = PCA9422_LDO3_OUT,
+	.ad_mask = PCA9422_LDO3_OUT_L3_AD_MASK,
+	.ad_pos = PCA9422_LDO3_OUT_L3_AD_POS,
+
+	.vsel_mode = {
+		[0] = {
+			.vsel_reg = PCA9422_LDO3_OUT,
+			.vsel_mask = PCA9422_LDO3_OUT_L3_OUT_MASK,
+		},
+		[1] = {
+			.vsel_reg = PCA9422_LDO3_OUT_SLEEP,
+			.vsel_mask = PCA9422_LDO3_OUT_SLEEP_L3_OUT_SLEEP_MASK,
+		},
+		[2] = {
+			.vsel_reg = PCA9422_LDO3_OUT_STBY,
+			.vsel_mask = PCA9422_LDO3_OUT_STBY_L3_OUT_STBY_MASK,
+		},
+	},
+	.max_ua = 200000,
+	.ranges = ldo23_ranges,
+	.num_ranges = ARRAY_SIZE(ldo23_ranges),
+};
+
+static const struct regulator_pca9422_desc ldo4_desc = {
+	.enable_reg = PCA9422_REGULATOR_EN,
+	.enable_mask = PCA9422_REGULATOR_EN_L4_ENABLE_MASK,
+
+	.mode_reg = PCA9422_LDO4_CFG,
+	.enmode_mask = PCA9422_LDO4_CFG_L4_ENMODE_MASK,
+	.enmode_pos = PCA9422_LDO4_CFG_L4_ENMODE_POS,
+	/* LDO4 has no lpmode control */
+	.lpmode_mask = 0,
+	.lpmode_pos = 0,
+
+	.ad_reg = PCA9422_LDO4_CFG,
+	.ad_mask = PCA9422_LDO4_CFG_L4_AD_MASK,
+	.ad_pos = PCA9422_LDO4_CFG_L4_AD_POS,
+
+	.vsel_mode = {
+		[0] = {
+			.vsel_reg = PCA9422_LDO4_OUT,
+			.vsel_mask = PCA9422_LDO4_OUT_L4_OUT_MASK,
+		},
+		[1] = {
+			.vsel_reg = PCA9422_LDO4_OUT_SLEEP,
+			.vsel_mask = PCA9422_LDO4_OUT_SLEEP_L4_OUT_SLEEP_MASK,
+		},
+		[2] = {
+			.vsel_reg = PCA9422_LDO4_OUT_STBY,
+			.vsel_mask = PCA9422_LDO4_OUT_STBY_L4_OUT_STBY_MASK,
+		},
+	},
+	.max_ua = 200000,
+	.ranges = ldo4_ranges,
+	.num_ranges = ARRAY_SIZE(ldo4_ranges),
+};
+
+static unsigned int regulator_pca9422_count_voltages(const struct device *dev)
+{
+	const struct regulator_pca9422_config *config = dev->config;
+
+	return linear_range_group_values_count(config->desc->ranges,
+					       config->desc->num_ranges);
+}
+
+static int regulator_pca9422_list_voltage(const struct device *dev,
+					  unsigned int idx, int32_t *volt_uv)
+{
+	const struct regulator_pca9422_config *config = dev->config;
+
+	return linear_range_group_get_value(config->desc->ranges,
+					    config->desc->num_ranges, idx,
+					    volt_uv);
+}
+
+static int regulator_pca9422_lock(const struct i2c_dt_spec *spec)
+{
+	return i2c_reg_write_byte_dt(spec, PCA9422_REG_LOCK, PCA9422_REG_LOCK_LOCK);
+}
+
+static int regulator_pca9422_unlock(const struct i2c_dt_spec *spec)
+{
+	return i2c_reg_write_byte_dt(spec, PCA9422_REG_LOCK, PCA9422_REG_LOCK_UNLOCK);
+}
+
+static int regulator_pca9422_update_locked_reg(const struct i2c_dt_spec *spec,
+					       uint8_t reg_addr, uint8_t mask, uint8_t value)
+{
+	int ret;
+
+	/* have to unlock first or else register read returns 0x0 */
+	ret = regulator_pca9422_unlock(spec);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* do the update while unlocked */
+	ret = i2c_reg_update_byte(spec->bus, spec->addr, reg_addr, mask, value);
+
+	/* always unlock and return result of the update */
+	regulator_pca9422_lock(spec);
```

Comment:
lock command is also an I2C write which may fail, so needs a ret check here.
It would be OK to return after the update byte if it fails, as in that case the lock command will probably fail too.

Reply:
Even if the update fails, I thought it better to make an attempt to restore the lock state.